File: src/DashSpec.Core/Parsing/LayoutModuleParser.cs
using DashSpec.Core.Model;

namespace DashSpec.Core.Parsing;

/// <summary>Transitional entry — delegates to F# Modeling.Parse via <see cref="LayoutParseBridge"/>.</summary>
public static class LayoutModuleParser
{
    public static LayoutBoardDefinition ParseLayoutFile(string text)
    {
        ArgumentException.ThrowIfNullOrWhiteSpace(text);
        if (LayoutParseBridge.ParseLayoutFile is { } parse)
        {
            return parse(text);
        }

        throw new InvalidOperationException(
            "Layout parse bridge not registered. Reference DashSpec.Execution.Core or register LayoutParseBridge.ParseLayoutFile.");
    }

    public static LayoutBoardDefinition Load(string reference, string specDirectory)
    {
        ArgumentException.ThrowIfNullOrWhiteSpace(reference);
        ArgumentException.ThrowIfNullOrWhiteSpace(specDirectory);

        var path = SpecIncludeResolver.ResolvePath(reference, specDirectory);
        path = ResolveLayoutFile(path);

        if (!File.Exists(path))
        {
            throw new FileNotFoundException(
                $"Include layout not found: '{reference}' (resolved: {path}).",
                path);
        }

        return ParseLayoutFile(File.ReadAllText(path));
    }

    private static string ResolveLayoutFile(string path)
    {
        if (File.Exists(path))
        {
            return path;
        }

        const string extension = ".dashlayout";
        var withExt = path.EndsWith(extension, StringComparison.OrdinalIgnoreCase) ? path : path + extension;
        return File.Exists(withExt) ? withExt : path;
    }
}

File: src/DashSpec.Modeling.Parse/Layout/LayoutModuleParser.fs
namespace DashSpec.Modeling.Parse.Layout

open DashSpec.Core.Parsing

/// <summary>
/// M2 pilot entry — delegates lex+parse to canonical C# <see cref="LayoutModuleParser"/>.
/// F# owns IR mapping; no duplicate token layer until Modeling lex spine lands (M4+).
/// </summary>
module LayoutModuleParser =

    let private mapScope (scope: DashSpec.Core.Model.LayoutScope) =
        match scope with
        | DashSpec.Core.Model.LayoutScope.Toolbar -> LayoutScope.Toolbar
        | DashSpec.Core.Model.LayoutScope.Tab -> LayoutScope.Tab
        | DashSpec.Core.Model.LayoutScope.Page -> LayoutScope.Page
        | DashSpec.Core.Model.LayoutScope.Card -> LayoutScope.Card
        | _ -> LayoutScope.Toolbar

    let private fromCore (board: DashSpec.Core.Model.LayoutBoardDefinition) : LayoutBoardDefinition =
        { Rows = board.Rows
          ModuleScope = board.ModuleScope |> Option.ofNullable |> Option.map mapScope }

    /// <summary>Parse a <c>.dashlayout</c> file body via transitional C# parser.</summary>
    let parseLayoutFile (text: string) : LayoutBoardDefinition =
        text
        |> DashSpec.Core.Parsing.LayoutModuleParser.ParseLayoutFile
        |> fromCore
